### ISSUE TITLE
[gatsby-remark-prismjs] `diff` with language / highlight

### ISSUE THREAD

**Issue by stephencweiss** [gatsbyjs/gatsby#20630]
Hi there! 

Trying to understand a few things related to `gatsby-remark-prismjs` and how it may (or may not) differ from regular `prismjs`.

The prism site has an example of [diff-highlighting](https://prismjs.com/plugins/diff-highlight/#example) and I'm trying to recreate that in my site ([here's my post](https://github.com/stephencweiss/stephencharlesweiss.com/blob/new-posts-week-20200127/content/blog/2020-01-15/syntax-highlighting-prismjs-gatsby.md#diff-highlighting)).

Github's preview is a good example of what I'm striving for:
![Github Preview](https://user-images.githubusercontent.com/39878535/72468530-ff8ece00-37a2-11ea-97c7-f8ff0a744a56.png)

On my site, however, I can't seem to figure out how to get highlights - whether language specific or not:
![Screen Shot 2020-01-15 at 2 26 31 PM](https://user-images.githubusercontent.com/39878535/72468553-087f9f80-37a3-11ea-9e75-f992e2f13189.png)

I've tried 
````
```diff-javascript
```diff javascript
````
And several other variants, but to no avail. 

Interestingly, `diff-javascript` results in some weird encoding (note the `undefined` values that get appended to the ends of lines):

![Screen Shot 2020-01-15 at 2 29 23 PM](https://user-images.githubusercontent.com/39878535/72468720-6e6c2700-37a3-11ea-9a27-08ca623dbc38.png)

My questions:
1. Does `gatsby-remark-prismjs` support multi-language (diff + javascript for example)?
2. Does `gatsby-remark-prismjs` support line-highlighting (with the green/red, not the standard highlighting)?

Thanks!

**Comment by sidharthachatterjee** [comment]
@stephencweiss Have you tried https://www.gatsbyjs.org/packages/gatsby-remark-prismjs/#optional-add-line-highlighting-styles?

**Comment by stephencweiss** [comment]
@sidharthachatterjee Yes, I believe so. 

Reading over my initial post, I may not have been totally clear in my original question. Let me try again. 

I have highlighting configured on my site (see the screen shot below). 

What I'm missing is what the CSS would look like for the `diff` language so that it's red for `-` lines and green `+` lines. Per [Prism's site, this appears to be called `diff-highlight`](https://prismjs.com/plugins/diff-highlight/#how-to-use).

For example, right now, my best solution is to _manually_ highlight the diff rows which is not ideal for two reasons.
1. It's cumbersome (and I think it should be included in the prismjs plugin
2. It doesn't differentiate between new lines (`+`) and deleted ones (`-`)

For example, typing this:
````
regular diff
```diff
{
-    resolve: `old-plugin`,
+    resolve: `gatsby-remark-prismjs`,
    options: {
        inlineCodeMarker: '>',
    },
}
```

regular diff with (standard) highlighting
```diff
{
-    resolve: `old-plugin`, // highlight-line
+    resolve: `gatsby-remark-prismjs`, // highlight-line
    options: {
        inlineCodeMarker: '>',
    },
}
```
````
Looks like this in my rendering:
![Screen Shot 2020-01-23 at 12 39 55 PM](https://user-images.githubusercontent.com/39878535/73013227-882def80-3ddd-11ea-9592-81ce6d803444.png)

My understanding from reading the [Prism docs](https://prismjs.com/plugins/diff-highlight/#example) is there should be a way to apply the class: `class="language-diff-javascript diff-highlight"`. 

This would:
1. Note that the languages are _both_ `diff` and `javascript`
2. Highlight rows with the red/green automatically

Honestly though, the class I _really_ want is just `"language-diff diff-highlight`

Thanks!

Below are how the different classes look per the Prism docs:
![Screen Shot 2020-01-23 at 12 44 56 PM](https://user-images.githubusercontent.com/39878535/73013586-3afe4d80-3dde-11ea-8e56-89049cb0c13b.png)

**Comment by stephencweiss** [comment]
Said another way:
1. Is `diff-highlighting` built into the `remark-prism` plugin?
2. If it is, is there a way to indicate within the code fence that you want to apply that style? Similar to how you add a title by saying:
````
```lang:title="my title"
  //code goes here
```
````

**Comment by stephencweiss** [comment]
> Hiya!
> 
> This issue has gone quiet. Spooky quiet. 👻
> 
> We get a lot of issues, so we currently close issues after 30 days of inactivity. It’s been at least 20 days since the last update here.
> If we missed this issue or if you want to keep it open, please reply here. You can also add the label "not stale" to keep this issue open!
> As a friendly reminder: the best way to see this issue, or any other, fixed is to open a Pull Request. Check out [gatsby.dev/contribute](https://www.gatsbyjs.org/contributing/how-to-contribute/) for more information about opening PRs, triaging issues, and contributing!
> 
> Thanks for being a part of the Gatsby community! 💪💜

Would love some help on this as I still haven't solved it :-/

**Comment by jamessimone** [comment]
@stephencweiss did you ever figure out anything to address this?

**Comment by stephencweiss** [comment]
I did not. It’s on my list to switch over to mdx and some more standard styling to see if that helps, but haven’t had a chance to test it.

Stephen Weiss | 651.233.6827

**Comment by jon-sully** [comment]
Greetings 👋🏻

This bothered me too, especially since I _don't_ use the line-highlighting implemented by this plugin. I modded a bit of the diff-sniffing code behind this plugin and published a variant that injects the `diff-highlight` class. I likely won't PR this repo with the changes because they're not the cleanest and it sort of goes against Commonmark spec, but you're welcome to use it if you'd like:

(before/after)
<img width="371" alt="image" src="https://user-images.githubusercontent.com/8585984/115182030-3428be00-a0a7-11eb-9660-1349b538c26e.png">


```shell
yarn remove gatsby-remark-prismjs
yarn add @jonsully/gatsby-remark-prismjs
```

In `gatsby-config.js`:

```diff
- resolve: `gatsby-remark-prismjs`,
+ resolve: `@jonsully/gatsby-remark-prismjs`,
```

In `gatsby-browser.js`:

```diff
+ require("prismjs/plugins/diff-highlight/prism-diff-highlight.css")
```

**Usage** (from above example):

Where we previously have written (in markdown):

<pre>
<code>
```diff-js
  const foo = input => {
+   console.log(input)
-   return bar
  }
```
</code>
</pre>

We can now add a dot (.) and the diff-highlight directive:

<pre>
<code>
```diff-js.diff-highlight
  const foo = input => {
+   console.log(input)
-   return bar
  }
```
</code>
</pre>

Give it a try if ya like. Your mileage may vary 🙂

**Comment by jon-sully** [comment]
Lol. I've since woken up to a new day and realized how bad of a solution that was.

I'm going to whip up a new solution that's _much_ simpler and instead exposes a plugin-level config option like `showDiffHighlight` that just always adds the `diff-highlight` class when enabled (plugin-wide). I _will_ PR this plugin with that change because it feels _way_ cleaner.

Apologies for the toss-n-turn. @stephencweiss (and others) if you did pull down my variant, please switch off of it for the time being 😅 ultimately I'd like to yank it from NPM and give it a nice, swift death 😛

**Comment by jon-sully** [comment]
PR with that premise is opened at #30939.

Until the maintainers have time to see it / consider it / etc, I've released that feature in its own package at https://www.npmjs.com/package/@jonsully/gatsby-remark-prismjs. To install and use until the PR is merged and a new Gatsby release is cut, do the following (this is also in the README updates in the PR):


Run:
```shell
$> yarn remove gatsby-remark-prismjs
$> yarn add @jonsully/gatsby-remark-prismjs
```

In gatsby-config.js:

```diff
- resolve: `gatsby-remark-prismjs`,
+ resolve: `@jonsully/gatsby-remark-prismjs`,
```

In gatsby-browser.js:

```diff
+ require("prismjs/plugins/diff-highlight/prism-diff-highlight.css")
```

**Usage:**

Automatic — anywhere you have a code block using `diff-xyz_language` line highlighting should automatically work.

Give it a try if ya like. Your mileage may vary 🙂

**Comment by stephencweiss** [comment]
> Lol. I've since woken up to a new day and realized how bad of a solution that was.
> 
> I'm going to whip up a new solution that's _much_ simpler and instead exposes a plugin-level config option like `showDiffHighlight` that just always adds the `diff-highlight` class when enabled (plugin-wide). I _will_ PR this plugin with that change because it feels _way_ cleaner.
> 
> Apologies for the toss-n-turn. @stephencweiss (and others) if you did pull down my variant, please switch off of it for the time being 😅 ultimately I'd like to yank it from NPM and give it a nice, swift death 😛

Well, you wouldn't be coding something novel if you didn't come up with a better solution immediately after your first approach :) I stand by my original assessment: I appreciate the work!

Also, one use case that I'm curious about, would the following be parsed correctly (the title part):
````
```diff-javascript:title=foo.js
module.exports = {
-   foo: "baz"
+   foo: "bar"
}
```
````

**Comment by jon-sully** [comment]
That's a good question. I haven't used (or seen) that syntax before — what are you attempting to do with it? And does it currently work with the plugin? The diffs I made shouldn't break any existing functionality at all.

**Comment by stephencweiss** [comment]
Yup! https://stephencharlesweiss.com/styling-markdown-code-snippets-with-prism-js#code-titles

**Comment by jon-sully** [comment]
Oh I see — this is coming from a separate, additional plugin.

https://github.com/DSchau/gatsby-remark-code-titles

I checked the code of that plugin just _briefly_ and I don't think there's anything that should necessarily change how it works with `gatsby-remark-prismjs`. If it works for you now it should continue to without issue.

**Comment by AurelianSpodarec** [comment]
Quick question, how about doing this in React/Vite?


This is meant to work:?
````
diff-js:title="config.ts"
```

This is really confusing docs. 

So far only "diff" works, but any other combindation with diff-js or js-diff does not. 

Let alone a title.
